staging/src/k8s.io/kubectl/pkg/cmd/create/create_secret.go

```diff
@@ -283,6 +283,7 @@ func NewCmdCreateSecretTLS(f cmdutil.Factory, ioStreams genericclioptions.IOStre
 	cmdutil.AddGeneratorFlags(cmd, generateversioned.SecretForTLSV1GeneratorName)
 	cmd.Flags().String("cert", "", i18n.T("Path to PEM encoded public key certificate."))
 	cmd.Flags().String("key", "", i18n.T("Path to private key associated with given certificate."))
+	cmd.Flags().String("ca", "", i18n.T("Path to CA certificate bundle associated with given certificate."))
```

Comment:
The typical requirement for a CA bundle is not that it is associated with the certificate being presented, but that it is valid for the certificates presented by the remote end of the TLS connection. The relationship between the CA bundle and the cert/key is unclear to me.

Reply:
Typically, when a CA issues a server certificate, they don't issue it from the trusted root, but from an intermediate CA that they can invalidate if necessary. Client applications don't have the intermediate CA certificates in their trust stores, but only the public root. 

When a CA issues a certificate to a user, the CA will provide both the leaf server certificate, and a bundle of intermediate certificates to establish the chains of trust to the trusted root. This bundle is the CA bundle that is invariably needed when setting up a TLS server.

So when you use a Secret to store a TLS certificate for Ingress, you also need some way to specify the CA bundle. cert-manager does it using the `ca.crt` key and a number of Ingress controllers support this scheme due to the popularity of cert-manager.

https://letsencrypt.org/certificates/
https://www.sslmarket.com/ssl/intermediate-and-root-certificates-ca-bundle/

In the Let's Encrypt case, on macOS I only have the "ISRG Root X1" certificate in my system keychain, but they actually issue certificates from "Let’s Encrypt Authority X3". This means that any server using a Let's Encrypt certificate must be configured with the CA bundle so that the server can send the intermediate certificates and the client can establish the chain of trust.

Reply:
Intermediates are often appended to the certificate content itself. The common uses of ca.crt I've seen are to specify a bundle to use to verify the other actor, not as a bundle of intermediates to present. Mixing up the two uses would lead to verification bugs.

Reply:
Appending to certificate content is what I have seen as well for handling intermediates.

(repeating what was said above)
Isn't ca.crt usually the trusted root?

Reply:
The reason we support the CA bundle (with any combination of root and intermediate) certificates in Contour is because cert-manager generates them. IIUC, cert-manager generates them because of ingress-nginx supports https://github.com/jetstack/cert-manager/pull/848. It's not immediately obvious to me whether the Istio support in https://github.com/istio/istio/pull/22622  is for client or server certificates.

I haven't seen this field used for specifying the CA to validate a TLS peer against, and from my perspective, it seems a bit unnatural to bundle that with a client certificate, since (at least in principle), it might be different for each peer.

How would you recommend we proceed here?

1. Drop the PR?
2. Drop the just kubectl changes?
3. Update the doc comment for the key constant to be neutral WRT how to use the key?

Reply:
> I haven't seen this field used for specifying the CA to validate a TLS peer

IIUC, that's exactly what ingress-nginx does, see https://kubernetes.github.io/ingress-nginx/examples/auth/client-certs/#creating-certificate-secrets:

> You can create a secret containing CA certificate along with the Server Certificate, that can be used for both TLS and Client Auth.
>
>> kubectl create secret generic ca-secret --from-file=tls.crt=server.crt --from-file=tls.key=server.key --from-file=ca.crt=ca.crt

tls.crt/tls.key are the presented serving certificate

ca.crt is the CA bundle used to verify the TLS peer certificates

> How would you recommend we proceed here?

If we want to identify recommended keys for uses, I think we should be as clear as possible, not neutral or ambiguous about the meaning of a file that can cause significant security problems if used for an unexpected purpose.

For example, if we needed to recommend a separate filename to contain intermediate certs to present, a name like `intermediates.crt` would be much clearer (though I still am more familiar with the approach of appending intermediates to the certificate file).

Reply:
Hmm, in that case I'm probably not the right person to take this forward. The use of ca.crt as a peer validation root isn't a pattern that I'm familiar with or really feel comfortable advocating here.

Thanks for reviewing and taking the time to discuss this :)